README
dsari - Do Something and Record It
==================================

dsari is a lightweight continuous integration (CI) system. It provides
scheduling, concurrency management and trigger capabilities, and is easy
to configure. Job scheduling is handled via dsari-daemon, while
dsari-render may be used to format job run information as HTML.

Requirements
------------

dsari requires Python 2.6 or later, and will run on Unix-based
platforms. It requires the following non-core modules:

-   croniter, for parsing cron-style schedule definitions
-   Jinja2, for rendering HTML reports

Installation
------------

dsari may be installed as any normal Python package:

    $ sudo python setup.py install

When this is done, dsari will expect its configuration file --
dsari.json -- in /usr/local/etc/dsari/, and will store its data in
/usr/local/lib/dsari/.

These locations may be customized by passing -c argument to dsari-* to
specify the configuration directory, and the data_dir configuration
option, respectively.

When dsari is installed directly in /usr/ (i.e. as part of distribution
packaging), the default configuration and data directories will be
/etc/dsari/ and /var/lib/dsari/, respectively.

dsari does not need to be installed at all, it can be run directly from
the repository directory. In this case, the default configuration and
data directories will be ~/.dsari/etc/ and ~/.dsari/var/, respectively.

The rest of these documents assume a locally-running setup, i.e.
~/.dsari/.

Configuration
-------------

A basic configuration for dsari.json looks as follows:

    {
        "jobs": {
            "sample-job": {
                "command": ["/usr/bin/env"],
                "schedule": "H/5 * * * *"
            }
        }
    }

This defines a job named "sample-job", which is run every 5 minutes.
Many more configuration options are available in the doc/ directory.

Running
-------

Once dsari is configured, run dsari-daemon. By default, dsari-daemon
will run in the foreground, and can be used with a supervisor (upstart,
systemd, supervisord, etc). If given -d, it will daemonize.

When a job is scheduled to be run, it produces a "run". Runs are
identified by a UUID, the run output is stored in ~/.dsari/var/runs/,
and data related to the run (start time, stop time, exit code, etc) is
stored in a SQLite database at ~/.dsari/var/dsari.sqlite3.

When a run is executed, several environment variables are passed to the
program to be run:

    JOB_NAME=sample-job
    RUN_ID=fa0490b8-7a8e-4f6b-b73c-160199a9ff75
    PREVIOUS_RUN_ID=e5bd61b3-27f3-46ca-8169-372433056fc2
    PREVIOUS_SCHEDULE_TIME=1437004689.27
    PREVIOUS_START_TIME=1437004689.65
    PREVIOUS_STOP_TIME=1437004689.71
    PREVIOUS_EXIT_CODE=0

PREVIOUS_* variables are not set if there is no previous run. In
addition, several extra environment variables are present, if the job's
jenkins_environment option is set, to aid with migrations from Jenkins
setups:

    BUILD_NUMBER=fa0490b8-7a8e-4f6b-b73c-160199a9ff75
    BUILD_ID=fa0490b8-7a8e-4f6b-b73c-160199a9ff75
    BUILD_URL=file:///home/user/.dsari/var/runs/sample-job/fa0490b8-7a8e-4f6b-b73c-160199a9ff75/
    NODE_NAME=master
    BUILD_TAG=dsari-sample-job-fa0490b8-7a8e-4f6b-b73c-160199a9ff75
    JENKINS_URL=file:///home/user/.dsari/var/
    EXECUTOR_NUMBER=0
    WORKSPACE=/home/user/.dsari/var/runs/sample-job/fa0490b8-7a8e-4f6b-b73c-160199a9ff75

Reports
-------

To render HTML reports, run dsari-render occasionally. This will produce
a series of HTML files in ~/.dsari/var/html/. You may then serve these
files, rsync them to a remote server, etc.

License
-------

dsari - Do Something and Record It

Copyright (C) 2015 Ryan Finnie

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.


HISTORY.md
# History

dsari was written to scratch a personal itch:
[I](http://www.finnie.org/) wanted a lightweight CI environment for [Finnix](http://www.finnix.org/) builds, but had grown beyond what simple cron scripts could provide.
I had been using Jenkins, but found the Java and memory requirements to be excessive, especially on armhf builders.
So I went and began writing my own scheduler, but generally didn't want to take myself too seriously.

Preserved below is the original README.md, written before any code had been produced.
Enjoy!

# dsari - Do Something and Record It

## About
dsari is described as a "Jenkins killer", in that I want to kill Jenkins.  It is a small scheduler which does something (something you configure) and records the output (somewhere).

## Development goals
1. Write a small script which does 85% of the features I need, and a random handful of features I don't really need.
2. Deploy it and forget about it for the next year or so.
3. Add a handful of extra features.
4. Realize six other people are also using dsari, two of which have filed a combined 76 bugs and feature requests.
  1. Fix three of the bugs.
  2. Implement two of the feature requests.
  3. Mark the other 71 WONTFIX.
5. Add a bunch more functionality, making the entire project unwieldy and complex.
6. Rename the project after a C&D from a commercial project with a similar name.
7. Rewrite the entire project in Go.
8. Abandon the project.
9. Sourceforge picks up the project files, wraps them in malware and re-releases them.


dsari/daemon.py
#!/usr/bin/env python

# dsari - Do Something and Record It
# Copyright (C) 2015 Ryan Finnie
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.

import os
import sys
import json
import time
import uuid
import math
import random
import logging
import signal
import sqlite3
import argparse
import copy
import re
import binascii
import __init__ as dsari
import croniter_hash
import utils


def wait_deadline(pid, options, deadline, interval=0.05):
    while True:
        (child_pid, child_exit, child_resource) = os.wait4(pid, options)
        child_signal = child_exit % 256
        if child_signal > 0:
            child_exit = 128 + child_signal
        else:
            child_exit = child_exit >> 8
        if child_pid != 0:
            return (child_pid, child_exit, child_resource)
        if time.time() >= deadline:
            return (child_pid, child_exit, child_resource)
        time.sleep(interval)


def backoff(a, b, min=5.0, max=300.0):
    if a >= b:
        return min
    r = 2 ** math.log(b - a)
    if r < min:
        return min
    elif r > max:
        return max
    else:
        return r


def parse_args():
    parser = argparse.ArgumentParser(
        description='Do Something and Record It - scheduler daemon (%s)' % dsari.VERSION,
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
    )
    parser.add_argument(
        '--config-dir', '-c', type=str, default=utils.DEFAULT_CONFIG_DIR,
        help='configuration directory for dsari.json',
    )
    parser.add_argument(
        '--fork', action='store_true',
        help='fork into the background after starting',
    )
    parser.add_argument(
        '--debug', action='store_true',
        help='output additional debugging information',
    )
    return parser.parse_args()


class Job():
    def __init__(self, name):
        self.name = name
        self.config = {}


class Run():
    def __init__(self, job, id):
        self.job = job
        self.id = id
        self.schedule_time = None
        self.trigger_type = 'schedule'
        self.trigger_data = {}
        self.respawn = False


class Scheduler():
    def __init__(self, args):
        self.shutdown = False
        self.args = args
        self.load_config()
        for signum in (signal.SIGHUP, signal.SIGINT, signal.SIGTERM, signal.SIGQUIT):
            signal.signal(signum, self.signal_handler)

        self.logger = logging.getLogger()
        self.logger.setLevel(logging.DEBUG)
        lh_console = logging.StreamHandler()
        lh_console_formatter = logging.Formatter('[%(asctime)s] %(levelname)s: %(message)s')
        lh_console.setFormatter(lh_console_formatter)
        if self.args.debug:
            lh_console.setLevel(logging.DEBUG)
        else:
            lh_console.setLevel(logging.INFO)
        self.logger.addHandler(lh_console)

        if not os.path.exists(self.config['data_dir']):
            os.makedirs(self.config['data_dir'])
        db_exists = os.path.exists(os.path.join(self.config['data_dir'], 'dsari.sqlite3'))
        self.db_conn = sqlite3.connect(os.path.join(self.config['data_dir'], 'dsari.sqlite3'))
        if not db_exists:
            sql_statement = """
                CREATE TABLE runs (
                    job_name text,
                    run_id text,
                    schedule_time real,
                    start_time real,
                    stop_time real,
                    exit_code integer,
                    trigger_type text,
                    trigger_data text,
                    run_data text
                )
            """
            self.db_conn.execute(sql_statement)
            self.db_conn.commit()

        self.reset_jobs()

        self.running_runs = []
        self.running_groups = {}

        self.wakeups = []
        self.next_wakeup = time.time() + 60.0

        self.logger.info('Scheduler running')

    def begin_shutdown(self):
        for run in copy.copy(self.runs):
            if run not in self.running_runs:
                self.runs.remove(run)

        if self.config['shutdown_kill_runs']:
            for run in self.running_runs:
                if run.term_sent:
                    continue
                job = run.job
                self.logger.info('[%s %s] Shutdown requested, sending SIGTERM to %d' % (job.name, run.id, run.pid))
                os.kill(run.pid, signal.SIGTERM)
                run.term_sent = True
        elif len(self.running_runs) > 0:
            self.logger.info('Shutdown will proceed after runs have completed')

    def monitor_shutdown(self):
        if not self.config['shutdown_kill_runs']:
            return
        if not self.config['shutdown_kill_grace']:
            return
        if time.time() < (self.shutdown_begin + self.config['shutdown_kill_grace']):
            self.wakeups.append(self.shutdown_begin + self.config['shutdown_kill_grace'])
            return
        for run in self.running_runs:
            if run.kill_sent:
                continue
            job = run.job
            self.logger.info('[%s %s] Shutdown grace time exceeded, sending SIGKILL to %d' % (job.name, run.id, run.pid))
            os.kill(run.pid, signal.SIGKILL)
            run.kill_sent = True

    def signal_handler(self, signum, frame):
        if signum in (signal.SIGINT, signal.SIGTERM):
            if signum == signal.SIGINT:
                self.logger.info('SIGINT received, beginning shutdown')
            elif signum == signal.SIGTERM:
                self.logger.info('SIGTERM received, beginning shutdown')
            self.shutdown = True
            self.shutdown_begin = time.time()
            self.begin_shutdown()
        elif signum == signal.SIGHUP:
            self.logger.info('SIGHUP received, reloading config when next idle')
            self.sighup_load_config = True
        elif signum == signal.SIGQUIT:
            self.sigquit_status()

    def sigquit_status(self):
        now = time.time()
        for run in sorted(self.runs, key=lambda x: x.job.name):
            job = run.job
            if run in self.running_runs:
                t = run.start_time
                self.logger.info(
                    '[%s %s] PID %d running since %s (%0.02fs)' % (
                        job.name,
                        run.id,
                        run.pid,
                        time.strftime('%c', time.localtime(t)),
                        (now - t)
                    )
                )
            else:
                t = run.schedule_time
                self.logger.info(
                    '[%s %s] Next scheduled run: %s (%0.02fs)' % (
                        job.name,
                        run.id,
                        time.strftime('%c', time.localtime(t)),
                        (t - now)
                    )
                )

    def load_config(self):
        self.sighup_load_config = False
        self.config = utils.load_config(self.args.config_dir)

    def reset_jobs(self):
        if self.shutdown:
            return
        self.jobs = []
        self.runs = []
        now = time.time()
        for job_name in sorted(self.config['jobs']):
            if (len(job_name) > 64) or (not re.search('^([- A-Za-z0-9_+.:@]+)$', job_name)):
                self.logger.warning('Invalid job name: %s' % job_name)
                continue
            job = Job(job_name)
            job.config = self.config['jobs'][job_name]
            job.subsecond_offset = float(binascii.crc32(job.name) & 0xffffffff) / float(2**32)
            self.jobs.append(job)
            if not job.config['schedule']:
                self.logger.debug('[%s] No schedule defined, manual triggers only' % job.name)
                continue
            if len(job.config['schedule'].split(' ')) == 5:
                job.config['schedule'] = job.config['schedule'] + ' H'
            try:
                t = croniter_hash.croniter_hash(
                    job.config['schedule'],
                    start_time=now,
                    hash_id=job_name
                ).get_next() + job.subsecond_offset
            except Exception, e:
                self.logger.warning('[%s] Invalid schedule: %s: %s' % (job.name, type(e), str(e)))
                continue
            run = Run(job, str(uuid.uuid4()))
            run.respawn = True
            self.logger.debug(
                '[%s %s] Next scheduled run: %s (%0.02fs)' % (
                    job.name,
                    run.id,
                    time.strftime('%c', time.localtime(t)),
                    (t - now)
                )
            )
            run.schedule_time = t
            self.runs.append(run)

    def process_run_execution_time(self, run):
        job = run.job
        if not job.config['max_execution']:
            return
        sigterm_grace = job.config['max_execution_grace']
        sigkill_grace = 5.0
        now = time.time()
        delta = now - run.start_time
        if delta > (job.config['max_execution'] + sigterm_grace):
            if not run.kill_sent:
                self.logger.warning(
                    '[%s %s] SIGTERM grace (%0.02fs) exceeded, sending SIGKILL to %d' % (
                        job.name,
                        run.id,
                        sigterm_grace,
                        run.pid
                    )
                )
                os.kill(run.pid, signal.SIGKILL)
                run.kill_sent = True
            self.wakeups.append(now + sigkill_grace)
        elif delta > job.config['max_execution']:
            if not run.term_sent:
                self.logger.warn(
                    '[%s %s] Max execution (%0.02fs) exceeded, sending SIGTERM to %d' % (
                        job.name,
                        run.id,
                        job.config['max_execution'],
                        run.pid
                    )
                )
                os.kill(run.pid, signal.SIGTERM)
                run.term_sent = True
            self.wakeups.append(now + sigterm_grace)
        else:
            self.wakeups.append(run.start_time + job.config['max_execution'])

    def run_child_executor(self, run):
        # Reset all handled signals to default
        for signum in (signal.SIGHUP, signal.SIGINT, signal.SIGTERM, signal.SIGQUIT):
            signal.signal(signum, signal.SIG_DFL)

        # Put the child in its own process group to prevent SIGINT from
        # propagating to the children
        os.setpgid(os.getpid(), 0)

        # Close the database in the child
        self.db_conn.close()

        # Set environment variables
        job = run.job
        os.environ['JOB_NAME'] = job.name
        os.environ['RUN_ID'] = run.id
        if run.concurrency_group:
            os.environ['CONCURRENCY_GROUP'] = run.concurrency_group
        if run.previous_run:
            os.environ['PREVIOUS_RUN_ID'] = run.previous_run[0]
            os.environ['PREVIOUS_SCHEDULE_TIME'] = str(run.previous_run[1])
            os.environ['PREVIOUS_START_TIME'] = str(run.previous_run[2])
            os.environ['PREVIOUS_STOP_TIME'] = str(run.previous_run[3])
            os.environ['PREVIOUS_EXIT_CODE'] = str(run.previous_run[4])
        if job.config['jenkins_environment']:
            os.environ['BUILD_NUMBER'] = run.id
            os.environ['BUILD_ID'] = run.id
            os.environ['BUILD_URL'] = 'file://%s' % os.path.join(self.config['data_dir'], 'runs', job.name, run.id, '')
            os.environ['NODE_NAME'] = 'master'
            os.environ['BUILD_TAG'] = 'dsari-%s-%s' % (job.name, run.id)
            os.environ['JENKINS_URL'] = 'file://%s' % os.path.join(self.config['data_dir'], '')
            os.environ['EXECUTOR_NUMBER'] = '0'
            os.environ['WORKSPACE'] = os.path.join(self.config['data_dir'], 'runs', job.name, run.id)
        for (key, val) in job.config['environment'].items():
            os.environ[key] = str(val)
        if 'environment' in run.trigger_data and run.trigger_data['environment']:
            for (key, val) in run.trigger_data['environment'].items():
                os.environ[key] = str(val)

        # Set STDIN to /dev/null, and STDOUT/STDERR to the output file
        signal.signal(signal.SIGPIPE, signal.SIG_DFL)
        out_f = open(os.path.join(self.config['data_dir'], 'runs', job.name, run.id, 'output.txt'), 'w')
        devnull_f = open(os.devnull, 'r')
        os.dup2(devnull_f.fileno(), 0)
        os.dup2(out_f.fileno(), 1)
        os.dup2(out_f.fileno(), 2)
        devnull_f.close()
        out_f.close()

        # Build command line
        command = job.config['command']
        if job.config['command_append_run']:
            command.append(job.name)
            command.append(run.id)

        # chdir to the run directory
        os.chdir(os.path.join(self.config['data_dir'], 'runs', job.name, run.id))

        # Finally!
        os.execvp(command[0], command)

    def process_next_child(self):
        self.logger.debug('Waiting up to %0.02fs for running jobs' % (self.next_wakeup - time.time()))
        (child_pid, child_exit, child_resource) = wait_deadline(-1, os.WNOHANG, self.next_wakeup)
        if child_pid == 0:
            return child_pid
        run = None
        for r in self.running_runs:
            if r.pid == child_pid:
                run = r
                break
        if not run:
            return child_pid
        job = run.job
        now = time.time()
        schedule_time = run.schedule_time
        start_time = run.start_time
        stop_time = now
        self.logger.info('[%s %s] Finished with status %d in %0.02fs' % (job.name, run.id, child_exit, (stop_time - start_time)))
        sql_statement = """
            INSERT INTO runs (
                job_name,
                run_id,
                schedule_time,
                start_time,
                stop_time,
                exit_code,
                trigger_type,
                trigger_data,
                run_data
            ) VALUES (
                ?, ?, ?, ?, ?, ?, ?, ?, ?
            )
        """
        self.db_conn.execute(sql_statement, (
            job.name,
            run.id,
            schedule_time,
            start_time,
            stop_time,
            child_exit,
            run.trigger_type,
            json.dumps(run.trigger_data),
            json.dumps({})
        ))
        self.db_conn.commit()
        self.running_runs.remove(run)
        self.runs.remove(run)
        if run.concurrency_group and run in self.running_groups[run.concurrency_group]:
            self.running_groups[run.concurrency_group].remove(run)
        if (not self.shutdown) and run.respawn and job.config['schedule']:
            run = Run(job, str(uuid.uuid4()))
            run.respawn = True
            t = croniter_hash.croniter_hash(
                job.config['schedule'],
                start_time=now,
                hash_id=job.name
            ).get_next() + job.subsecond_offset
            run.schedule_time = t
            self.runs.append(run)
            self.logger.debug(
                '[%s %s] Next scheduled run: %s (%0.02fs)' % (
                    job.name,
                    run.id,
                    time.strftime('%c', time.localtime(t)), (t - now)
                )
            )
        return child_pid

    def process_triggers(self):
        if self.shutdown:
            return
        for job in self.jobs:
            self.process_trigger_job(job)

    def process_trigger_job(self, job):
        trigger_file = os.path.join(self.config['data_dir'], 'trigger', job.name, 'trigger.json')
        if not os.path.exists(trigger_file):
            return
        t = os.path.getmtime(trigger_file)
        try:
            f = open(trigger_file)
        except IOError, e:
            # Return silently, otherwise we spam the log during each loop
            return
        try:
            os.remove(trigger_file)
        except OSError, e:
            # Return silently, otherwise we spam the log during each loop
            f.close()
            return
        try:
            j = json.load(f)
        except ValueError, e:
            self.logger.error('[%s] Cannot load trigger: %s' % (job.name, e.message))
            f.close()
            return
        f.close()
        if type(j) != dict:
            self.logger.error('[%s] Cannot load trigger: Data must be a dict' % job.name)
            return
        self.logger.info('[%s] Trigger detected, creating trigger run' % job.name)
        run = Run(job, str(uuid.uuid4()))
        run.respawn = False
        run.trigger_type = 'file'
        run.trigger_data = j
        run.schedule_time = t
        self.runs.append(run)

    def process_run(self, run):
        now = time.time()
        job = run.job
        if run in self.running_runs:
            self.wakeups.append(now + backoff(run.schedule_time, now))
            return
        if run.schedule_time > now:
            self.wakeups.append(run.schedule_time)
            return
        run.concurrency_group = None
        if len(job.config['concurrency_groups']) > 0:
            job_concurrency_groups = copy.copy(job.config['concurrency_groups'])
            random.shuffle(job_concurrency_groups)
            for concurrency_group in job_concurrency_groups:
                if concurrency_group not in self.running_groups:
                    self.running_groups[concurrency_group] = []
                concurrency_inuse = len(self.running_groups[concurrency_group])
                if (
                    (concurrency_group in self.config['concurrency_groups'])
                    and ('max' in self.config['concurrency_groups'][concurrency_group])
                ):
                    concurrency_max = self.config['concurrency_groups'][concurrency_group]['max']
                else:
                    concurrency_max = 1
                if concurrency_inuse < concurrency_max:
                    run.concurrency_group = concurrency_group
                    break
            if not run.concurrency_group:
                self.wakeups.append(now + backoff(run.schedule_time, now))
                return

        sql_statement = """
            SELECT
                run_id,
                schedule_time,
                start_time,
                stop_time,
                exit_code
            FROM
                runs
            WHERE
                job_name = ?
            ORDER BY
                stop_time DESC
        """
        res = self.db_conn.execute(sql_statement, (job.name,))
        run.previous_run = res.fetchone()
        res.close()

        if not os.path.exists(os.path.join(self.config['data_dir'], 'runs', job.name, run.id)):
            os.makedirs(os.path.join(self.config['data_dir'], 'runs', job.name, run.id))

        self.logger.info('[%s %s] Running: %s' % (job.name, run.id, job.config['command']))
        run.start_time = now
        run.term_sent = False
        run.kill_sent = False
        child_pid = os.fork()
        if child_pid == 0:
            self.run_child_executor(run)
            raise OSError('run_child_executor returned, when it should not have')
        run.pid = child_pid
        self.running_runs.append(run)
        if run.concurrency_group:
            self.running_groups[run.concurrency_group].append(run)

    def process_wakeups(self):
        self.next_wakeup = time.time() + 60.0
        for wakeup in self.wakeups:
            if wakeup < self.next_wakeup:
                self.next_wakeup = wakeup

    def loop(self):
        while True:
            self.wakeups = []
            self.process_triggers()
            runs = copy.copy(self.runs)
            random.shuffle(runs)
            for run in runs:
                self.process_run(run)

            for run in self.running_runs:
                self.process_run_execution_time(run)

            if self.shutdown:
                self.monitor_shutdown()

            self.process_wakeups()

            if len(self.running_runs) > 0:
                while (len(self.running_runs) > 0) and (self.next_wakeup > time.time()):
                    if self.process_next_child() == 0:
                        break
            else:
                if self.shutdown:
                    self.logger.info('Shutdown complete')
                    return

                to_sleep = self.next_wakeup - time.time()
                if to_sleep > 0:
                    self.logger.debug('No running jobs, waiting %0.02fs' % to_sleep)
                    time.sleep(to_sleep)

                if self.sighup_load_config:
                    self.logger.info('Reloading config')
                    self.load_config()
                    self.reset_jobs()


def main(argv):
    args = parse_args()
    if args.fork:
        child_pid = os.fork()
        if child_pid > 0:
            return
    s = Scheduler(args)
    s.loop()


if __name__ == '__main__':
    sys.exit(main(sys.argv))
